packages/block-editor/src/hooks/style.js

```diff
@@ -0,0 +1,145 @@
+/**
+ * External dependencies
+ */
+import { mapKeys, kebabCase, isObject, entries } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { addFilter } from '@wordpress/hooks';
+import { hasBlockSupport } from '@wordpress/blocks';
+
+/**
+ * Internal dependencies
+ */
+import { COLOR_SUPPORT_KEY } from './color';
+
+const styleSupportKeys = [ COLOR_SUPPORT_KEY ];
+
+const hasStyleSupport = ( blockType ) =>
+	styleSupportKeys.some( ( key ) => hasBlockSupport( blockType, key ) );
+
+/**
+ * Flatten a nested Global styles config and generates the corresponding
+ * flattened CSS variables.
+ *
+ * @param  {Object} styles Styles configuration
+ * @return {Object}        Flattened CSS variables declaration
+ */
+function getCSSVariables( styles = {} ) {
+	const prefix = '--wp';
+	const token = '--';
+	const getNestedCSSVariables = ( config ) => {
+		let result = {};
+		entries( config ).forEach( ( [ key, value ] ) => {
+			if ( ! isObject( value ) ) {
+				result[ key ] = value;
+				return;
+			}
+
+			result = {
+				...result,
+				...mapKeys(
+					getNestedCSSVariables( value ),
+					( _, subkey ) => kebabCase( key ) + token + subkey
+				),
+			};
+		} );
```

Comment:
We can use vanilla JS for the for loop, which is about 2-3% faster according to my testing:
```suggestion
		for( const [ key, value ] of Object.entries( config ) ) {
			if ( ! isObject( value ) ) {
				result[ key ] = value;
				return;
			}

			result = {
				...result,
				...mapKeys(
					getNestedCSSVariables( value ),
					( _, subkey ) => kebabCase( key ) + token + subkey
				),
			};
		}
```

Reply:
lodash handles the cases where the input could be undefined so its less error prone.

Reply:
In my opinion, if `config` can be `undefined`/`null`, you should be handling that more explicitly via default parameters, nullish coalescing, and/or optional chaining. (Not to mention stricter type checking via something like JSDoc comments or TypeScript.) Invisibly, silently handling nullish cases is a recipe for unexpected behavior.

In this instance, either you should be prevented from calling `getNestedCSSVariables` with a `null`/`undefined` value by using stricter type checking (via JSDoc or TypeScript), or else a default parameter value should be used, e.g.:

```js
const getNestedCSSVariables = ( config = {} ) => {
```